packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_decl.hpp

```diff
@@ -324,158 +330,144 @@ namespace Thyra {
+          RCP<MueLuHalfXpOp> xpOp = rcp(new MueLuHalfXpOp(H));
+          xpPrecOp = rcp(new XpHalfPrecOp(xpOp));
 #else
-          TEUCHOS_TEST_FOR_EXCEPTION(true, MueLu::Exceptions::RuntimeError,
-                                     "Thyra::MueLuPreconditionerFactory: Tpetra does not support GO=int and or EpetraNode.");
+          TEUCHOS_TEST_FOR_EXCEPT(true);
 #endif
-        }
-#endif
-#if defined(HAVE_MUELU_EPETRA)// && defined(HAVE_MUELU_SERIAL)
-        if (bIsEpetra) {
-          RCP<ThyEpLinOp> epetr_precOp = rcp_dynamic_cast<ThyEpLinOp>(thyra_precOp);
-          RCP<MueEpOp>    muelu_precOp = rcp_dynamic_cast<MueEpOp>(epetr_precOp->epetra_op(),true);
-
-          H = rcp_dynamic_cast<MueLu::Hierarchy<Scalar,LocalOrdinal,GlobalOrdinal,Node> >(muelu_precOp->GetHierarchy());
-        }
-#endif
-        // TODO add the blocked matrix case here...
-
-        TEUCHOS_TEST_FOR_EXCEPTION(!H->GetNumLevels(), MueLu::Exceptions::RuntimeError,
-                                   "Thyra::MueLuPreconditionerFactory: Hierarchy has no levels in it");
-        TEUCHOS_TEST_FOR_EXCEPTION(!H->GetLevel(0)->IsAvailable("A"), MueLu::Exceptions::RuntimeError,
-                                   "Thyra::MueLuPreconditionerFactory: Hierarchy has no fine level operator");
-        RCP<MueLu::Level> level0 = H->GetLevel(0);
-        RCP<XpOp>    O0 = level0->Get<RCP<XpOp> >("A");
-        RCP<XpMat>   A0 = rcp_dynamic_cast<XpMat>(O0);
-
-        if (!A0.is_null()) {
-          // If a user provided a "number of equations" argument in a parameter list
-          // during the initial setup, we must honor that settings and reuse it for
-          // all consequent setups.
-          A->SetFixedBlockSize(A0->GetFixedBlockSize());
-        }
-
-        // set new matrix
-        level0->Set("A", A);
+        } else
+          {
+            const std::string userName = "user data";
+            Teuchos::ParameterList& userParamList = paramList.sublist(userName);
+            if (paramList.isType<RCP<XpmMV> >("Coordinates")) {
+              RCP<XpmMV> coords = paramList.get<RCP<XpmMV> >("Coordinates");
+              paramList.remove("Coordinates");
+              userParamList.set("Coordinates",coords);
+            }
+            if (paramList.isType<RCP<XpMV> >("Nullspace")) {
+              RCP<XpMV> nullspace = paramList.get<RCP<XpMV> >("Nullspace");
+              paramList.remove("Nullspace");
+              userParamList.set("Nullspace",nullspace);
+            }
+
+            // build a new MueLu RefMaxwell preconditioner
+            RCP<MueLu::Hierarchy<Scalar,LocalOrdinal,GlobalOrdinal,Node> > H = MueLu::CreateXpetraPreconditioner(A, paramList);
+            xpPrecOp = rcp(new MueLuXpOp(H));
+          }
 
-        H->SetupRe();
-      }
+      } else {
+        // reuse old MueLu hierarchy stored in MueLu Tpetra/Epetra operator and put in new matrix
```

Comment:
This comment is misleading ... should be only Tpetra, correct?

Reply:
No, I think that's correct at that point. We do support reuse for Epetra. The check for half precision happens below.